File:  Invert Binary Tree.java
// Time Complexity :
// Space Complexity :
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // int sum=0;
     public TreeNode invertTree(TreeNode root) {
    return  dfs(root);
     }
     public TreeNode dfs(TreeNode root){
         if(root==null){
             return null;
         }
         if(root.left!=null){
           root.left = dfs(root.left);   
         }
        if(root.right !=null){
         root.right = dfs(root.right);
        }
         
         TreeNode temp=root.left;
         root.left=root.right;
         root.right=temp;
         return root;
         
     }
 }

File: Invert_Binary_Tree.py
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None


// Your code here along with comments explaining your approach:
used BFS in this approach.
we travel level by level and for every node if we find left and right nodes for the given node then we reverse them and add them to the queue.

# BFS 
# Time complexity --> o(n)
# space complexity --> o(n)
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if root==None:
            return root
        queue=deque([root])
        while len(queue)!=0:
            node=queue.popleft()
            #Invert the left and right side of a node.
            node.left,node.right=node.right,node.left
            if node.left!=None:
                queue.append(node.left)
            if node.right!=None:
                queue.append(node.right)
        return root
 // Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None


// Your code here along with comments explaining your approach:
           
# DFS approach
# Time complexity --> o(n)
# space complexity --> o(height of the tree)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if root==None:
            return None
        left=self.invertTree(root.left)
        right=self.invertTree(root.right)
        #make the left side of the tree to the right and vice versa.
        root.right=left
        root.left=right
        return root
        
        
        